File: AdapterPattern/AdapterDemo.java
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

interface Communication {
  void sendMessage(String message);
}

class Email implements Communication {
  String recipient;

  Email(String recipient) {
    this.recipient = recipient;
  }

  public void sendMessage(String Message) {
    System.out.println(MessageFormat.format("Email sent to {0}", recipient));
  }
}

class SMS implements Communication {
  String recipient;

  SMS(String recipient) {
    this.recipient = recipient;
  }

  public void sendMessage(String Message) {
    System.out.println(MessageFormat.format("SMS Message sent to {0}", recipient));
  }
}

class Subscriber {

  String firstName;
  String lastName;
  List<Communication> notifier = new ArrayList<Communication>();

  Subscriber(String firstName, String lastName) {
    this.firstName = firstName;
    this.lastName = lastName;
  }

  public void addNotifier(Communication comm) {
    notifier.add(comm);
  }

  public void notifySubscriber(String message) {
    for (var comm : notifier)
      comm.sendMessage(message);
  }
}

public class AdapterDemo {

  public static void main(String[] args) {
    var list = new ArrayList<Subscriber>();
    var bill = new Subscriber("Bill", "Gates");
    bill.addNotifier(new Email("billg@microsoft.com"));
    list.add(bill);

    var elon = new Subscriber("Elon", "Musk");
    elon.addNotifier(new SMS("1-CALL-ELONMUSK"));
    list.add(elon);

    for (var person : list)
      person.notifySubscriber("Bill due in 3 days");
    System.out.println("Done");
  }
}

File: AdapterPattern/AdapterDemo.py
from abc import ABC, abstractmethod


class Communication(ABC):

    @abstractmethod
    def send_message(self, message: str):
        pass

class Email(Communication):
    def __init__(self, recipient: str):
        super().__init__()
        self.recipient: str = recipient

    def send_message(self, message: str):
        print(f"Email sent to {self.recipient}")

class SMS(Communication):
    def __init__(self, recipient: str):
        super().__init__()
        self.recipient: str = recipient

    def send_message(self, message: str):
        print(f"SMS sent to {self.recipient}")

class Subscriber:
    def __init__(self, first_name: str, last_name: str):
        self.first_name: str = first_name
        self.last_name: str = first_name
        self.notifier: list[Communication] = []

    def add_notifier(self, comm: Communication):
        self.notifier.append(comm)

    def notify_subscriber(self, message: str):
        for comm in self.notifier:
            comm.send_message(message)

person_list: list[Subscriber] = []
bill: Subscriber = Subscriber("Bill", "Gates")
bill.add_notifier(Email("billg@microsoft.com"))
person_list.append(bill)

elon: Subscriber = Subscriber("Elon", "Musk")
elon.add_notifier(SMS("1-CALL-ELONMUSK"))
person_list.append(elon)

for person in person_list:
    person.notify_subscriber("Bill due in 3 days")
